src/balance.rs

```diff
@@ -26,41 +42,186 @@ impl<T: Config> From<Error<T>> for StatusCode {
 
 /// Balance adapter for providing basic access to balance cheques within the MoveVM.
 #[derive(Clone)]
-pub struct BalanceAdapter<T: Config> {
+pub struct BalanceAdapter<T: Config + SysConfig>
+where
+    AccountIdOf<T>: core::cmp::Eq + core::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     _pd_config: PhantomData<T>,
+    /// Copy and virtual cheques record of involved users.
+    cheques: Rc<RefCell<HashMap<AccountIdOf<T>, BalanceOf<T>>>>,
+    /// Copy of initial state, without tracking it.
+    initial_state: HashMap<AccountIdOf<T>, BalanceOf<T>>,
 }
 
-impl<T: Config> BalanceAdapter<T> {
+impl<T: Config + SysConfig> BalanceAdapter<T>
+where
+    AccountIdOf<T>: core::cmp::Eq + core::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     /// Create a new [`BalanceAdapter`].
     pub fn new() -> Self {
         BalanceAdapter {
             _pd_config: PhantomData,
+            cheques: Rc::new(RefCell::new(HashMap::new())),
+            initial_state: HashMap::new(),
+        }
+    }
+
+    /// Inserts an initial balance state, e.g. the executor of a script.
+    pub fn write_cheque(
+        &mut self,
+        account: &AccountIdOf<T>,
+        balance: &BalanceOf<T>,
+    ) -> DispatchResult {
+        // TODO (neutrinoks): Enable this feature to ensure, that balance is available!
+        // self.ensure_can_withdraw(account, balance)?;
+
+        let mut cheques = self.cheques.borrow_mut();
+        if let Some(src_cheque) = cheques.get_mut(account) {
+            *src_cheque += *balance;
+        } else {
+            cheques.insert(account.clone(), *balance);
+        }
+
+        if let Some(src_cheque) = self.initial_state.get_mut(account) {
+            *src_cheque += *balance;
+        } else {
+            self.initial_state.insert(account.clone(), *balance);
+        }
+
+        Ok(())
+    }
+
+    /// Executes the true transactions on the blockchain/substrate side after execution of
+    /// Move-script.
+    pub fn apply_transactions(&self) -> DispatchResult {
+        use core::cmp::Ordering;
+
+        let zero = BalanceOf::<T>::zero();
+
+        self.cmp_with_initial_state()?;
+
+        let cheques = self.cheques.borrow();
+        let mut purse = NegativeImbalanceOf::<T>::zero();
+        let mut depts = Vec::<(&AccountIdOf<T>, BalanceOf<T>)>::new();
+
+        // Calculate balance differences and withdraw negative ones from user's accounts.
```

Comment:
Seems like an interesting approach 👍 
We can huddle together about this tomorrow - I'm curious about this